File: zad2/Java/src/MatrixFillers/Matrix.java
package MatrixFillers;

import java.util.ArrayList;

public class Matrix {

    private ArrayList<Integer>[] matrix;
    private int size;

    public Matrix(int size) {
        matrix = new ArrayList[size];
        this.size = size;
        for (int i = 0; i < size; i++) {
            matrix[i] = new ArrayList<>(size);
            for (int j = 0; j < size; j++) {
                matrix[i].add(0);
            }
        }
    }

    public int get(int i, int j){
        return matrix[i].get(j);
    }
    public void set(int i, int j, int value){
        matrix[i].set(j, value);
    }
    public int size(){
        return size;
    }

    public Matrix clone()
    {
        Matrix m = new Matrix(size);
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                m.set(i, j, matrix[i].get(j));
            }
        }
        return m;
    }

    private void validateIndicies(int row, int col)
    {
        if (row < 0 || col < 0 || row >= size || col >= size)
        {
            throw new IndexOutOfBoundsException("Invalid matrix range.");
        }
    }

    public static Matrix multiply(Matrix m1, Matrix m2){
        if (m1.size != m2.size) throw new IllegalArgumentException("Matrix sizes don't match.");

        Matrix mr = new Matrix(m1.size);

        for (int i = 0; i < m1.size; i++)
        {
            for (int j = 0; j <  m1.size; j++)
            {
                mr.set(i, j, calculateCell(m1, m2, i, j));
            }
        }
        return mr;
    }

    private static int calculateCell(Matrix m1, Matrix m2, int row, int col)
    {
        int res = 0;
        for (int i = 0; i < m1.size; i++)
        {
            res += m1.get(row, i) * m2.get(i, col);
        }

        return res;
    }
}

File: zad2/CS/PRV_zad2_CS/Matrix.cs
using System.Numerics;

namespace PRV2;

public class Matrix
{
    private List<List<int>> _matrix;
    public int Size { get; private set; }

    public Matrix(int size)
    {
        Size = size;
        _matrix = new List<List<int>>(size);
        for (int i = 0; i < size; i++)
        {
            _matrix.Add(new List<int>(size));
            for (int j = 0; j < size; j++)
            {
                _matrix[i].Add(0);
            }
        }
    }

    public Matrix Clone()
    {
        Matrix m = new Matrix(Size);
        for (int i = 0; i < Size; i++)
        {
            for (int j = 0; j < Size; j++)
            {
                m[i, j] = _matrix[i][j];
            }
        }
        return m;
    }
    
    private void ValidateIndicies(int row, int col)
    {
        if (row < 0 || col < 0 || row >= Size || col >= Size)
        {
            throw new IndexOutOfRangeException("Invalid matrix range.");
        }
    }

    public Matrix MultiplyWith(Matrix m) 
    {
        if (this.Size != m.Size) throw new ArgumentException("Matrix sizes don't match.");
        
        Matrix mr = new Matrix(this.Size);

        for (int i = 0; i < this.Size; i++)
        {
            for (int j = 0; j < this.Size; j++)
            {
                mr[i, j] = CalculateCell(m, i, j);
            }
        }
        return mr;
    }

    private int CalculateCell(Matrix m2, int row, int col)
    {
        int res = 0;
        for (int i = 0; i < this.Size; i++)
        {
             res += m2[i, col] * this[row, i];
        } 
        
        return res;
    }
    public int this[int row, int col]
    {
        get
        {
            ValidateIndicies(row, col);
            return _matrix[row][col];
        }
        set
        {
            ValidateIndicies(row, col);
            _matrix[row][col] = value;
        }
    }
}